### PR TITLE
Updates to SQL Server Plugin, better Azure Managed Instance support + more

### LINKED ISSUE #4222: Unable to use SQL Database in Azure with sqlserver plugin
### Relevant telegraf.conf:

[[inputs.sqlserver]]
   servers = [
     "Server=somesqlserver.database.windows.net;Port=1433;Database=db01-example;User ID=telegraf;Password=***;app name=telegraf;log=1"
   ]
   query_version = 2
   azuredb = true

OR:
   servers = [
     "Server=somesqlserver.database.windows.net;Port=1433;User ID=telegraf;Password=***;app name=telegraf;log=1"
   ]

### System info:

Telegraf v1.6.2 (git: release-1.6 1fb4283f), Windows Server 2016
Azure SQL Database with elastic pool and multiple databases:

az sql server create --name sql01 --resource-group rgsql --location westeurope --admin-user sa --admin-password ***
 --assign-identity
az sql elastic-pool create --server sql01 --resource-group rgsql --name sql01ep01 --edition Standard --dtu 100
az sql db create --server sql01 --resource-group rgsql --elastic-pool sql01ep01 --name db01-example

### Steps to reproduce:

I'm unable to configure Telegraf for Azure SQL Database as the grants differ from a classic SQL Server (where my configuration works). First attempt:

1. Create user as described in the docs and grant VIEW SERVER STATE and VIEW ANY DEFINITION.

This fails because there is no VIEW SERVER STATE in SQL Database. Neither is it possible to grant VIEW ANY DEFINITION. Fails with "Securable class 'server' not supported in this version of SQL Server".

Second attempt:

1. Connect to the server without specifying a database using the super user (sa) with telegraf -test.

This works, but (a) it only returns data for master and tempdb and not for db01-example and the other application databases, plus (b) it is not a good idea to use the super user for monitoring. I want to grant as few permissions as possible.

Third attempt:

1. Connect to the server and specify db01-example as database using the super user with telegraf -test.

This crashes with a nil pointer dereference and the log complains about VIEW DATABASE STATE permission denied in database master (apparently the super user can't do that when logged in to a specific database).

Forth attempt:

1. Create user telegraf in db01-example and grant VIEW DATABASE STATE and VIEW DEFINITION (in that database).
2. Connect to the server and specify db01-example using telegraf with telegraf -test.

This crashes with a nil pointer dereference. The log simply says that the connection was forcibly closed.

### Expected behavior:

The documentation should provide instructions on how to set up the telegraf database user for SQL Database as the documented permissions used with SQL Server don't exist. With the right permissions connecting as a user with access rights should return metrics for all databases in the server, not just master and tempdb, just like on a regular SQL Server. Alternatively the instructions should be to create one telegraf database user per database and that should work and return data for the given database.

### Actual behavior:

No configuration I have tried manages to return the expected metrics. Worst case there is a nil pointer dereference and Telegraf crashes, best case metrics for the system databases are returned.

I think this is only a documentation problem, surely someone has tested this as there is support for Azure SQL Database in the plugin. However, try as I might I can't find the right configuration. Ideally I would like to connect with a single user and get statistics for all databases, but if that is not supported I can connect once per database. I need to know how to grant permissions, though.

### Additional info:

[Include gist of relevant config, logs, etc.]

### LINKED ISSUE #4597: Add counter [SQLServer:Databases Percent Log Used] to enable monitoring for T-log utilization
## Feature Request
Please add counter [SQLServer:Databases Percent Log Used] to enable monitoring for T-log utilization

Opening a feature request kicks off a discussion.

### Proposal:
Please add counter [SQLServer:Databases Percent Log Used] to enable monitoring for T-log utilization

### Current behavior:
Current dashboard monitors following counters
- SQLServer:Databases Log File(s) Size (KB)
- SQLServer:Databases Log File(s) Used Size (KB)

While these are good for getting trend of log usage, they aren't helpful in determining the % log used

### Desired behavior:
- Add counter [SQLServer:Databases Percent Log Used] that will easily expose the value of current T-log utilization and can be used to raise alwerts

### Use case: [Why is this important (helps with prioritizing requests)]
- Using Grafana dashboard for both monitoring (trending) and real-time alerting
- Right in absence of counter [SQLServer:Databases Percent Log Used], we are using a custom PowerShell script to fetch and alert on T-log utilization

### PR DESCRIPTION
**Pull request by m82labs**
closes #4222 
closes #4597

### Required for all PRs:
- [ x ] Signed [CLA](https://influxdata.com/community/cla/).
- [ x ] Associated README.md updated.
- [ x ] Has appropriate unit tests.

### Changes
- Altered database IO stats to include stats on a per-file basis
- Added a few new perf counters including `Background Writer pages/sec` which tracks indirect checkpoints
- Refactored some queries to be compatible with Azure Managed Instances
- Changed to the `denisenkom` SQL package, prior to the change this plugin would not work properly with managed instances
- Some perf counters now grab data per database instead of just for the "_Total" counter instance
- Moved some newer wait types into a proper category
- Added a few new wait types to the list of types to ignore
- Now filtering for wait types that have accumulated some amount of waits (100 ms)

The majority of these changes came from Denzil Ribeiro from Microsoft, so big thanks to Denzil!

### CODE REVIEW

**Comment by danielnelson** [comment]
I think we need to modify the Gopkg.toml by hand:
```
-  name = "github.com/zensqlmonitor/go-mssqldb"
+  name = "github.com/denisenkom/go-mssqldb"
```
After this run `dep ensure`, this should be all we need but also run `dep check` to make sure all is well.

**Comment by danielnelson** [comment]
I think we need `dep ensure` still.  Do you have a list of the issues this addresses?

**Comment by m82labs** [comment]
Yep. I will get the new rep files committed shortly. This is more of an
enhancement than something addressing current issues. After this is merged
in I plan on working on a few of the open issues.

**Comment by danielnelson** [comment]
Maybe it matches some feature requests.  Can you take a look at, being careful not to link yet, these issues: 4569, 4597, 4523, 4398, 4382

**Comment by m82labs** [comment]
@danielnelson just added a fix for 4597, 4382 was already closed out via https://github.com/influxdata/telegraf/pull/4571

**Comment by danielnelson** [comment]
Looks like that pr was for 4398, closed now.

File: plugins/inputs/sqlserver/sqlserver.go [hunk 1]
```diff
@@ -477,41 +551,41 @@ WHERE	(
...
 				'Disk Write IO Throttled/sec',
 				'Disk Write IO/sec',
-				'Used memory (KB)'
+				'Used memory (KB)',
+				'Forwarded Recs/sec',
+				'Background Writer pages/sec',
+				'Percent Log Used'
 			)
-		) OR  (
+		)  OR  (
```
Comment (glinton) [inline plugins/inputs/sqlserver/sqlserver.go]: un-needed extra spacing.

File: plugins/inputs/sqlserver/sqlserver.go [hunk 2]
```diff
@@ -358,43 +360,123 @@ vfs.num_of_bytes_read AS read_bytes,
...
+	sku NVARCHAR(64),
+	engine_edition SMALLINT,
+	hardware_type VARCHAR(15),
+	total_storage_mb BIGINT,
+	available_storage_mb BIGINT,
 	uptime INT
 )
 
-IF OBJECT_ID('master.sys.dm_os_sys_info') IS NOT NULL
+IF  OBJECT_ID('master.sys.dm_os_sys_info') IS NOT NULL
```
Comment (glinton) [inline plugins/inputs/sqlserver/sqlserver.go]: un-needed extra spacing

File: plugins/inputs/sqlserver/sqlserver.go [hunk 3]
```diff
@@ -1113,14 +1186,15 @@ ws.wait_type NOT IN (
...
 	N'WAIT_FOR_RESULTS', N'WAITFOR', N'WAITFOR_TASKSHUTDOWN', N'WAIT_XTP_HOST_WAIT',
 	N'WAIT_XTP_OFFLINE_CKPT_NEW_LOG', N'WAIT_XTP_CKPT_CLOSE',
 	N'XE_BUFFERMGR_ALLPROCESSED_EVENT', N'XE_DISPATCHER_JOIN',
-	N'XE_DISPATCHER_WAIT', N'XE_LIVE_TARGET_TVF', N'XE_TIMER_EVENT')
+	N'XE_DISPATCHER_WAIT', N'XE_LIVE_TARGET_TVF', N'XE_TIMER_EVENT',
+	N'SOS_WORK_DISPATCHER','RESERVED_MEMORY_ALLOCATION_EXT')
 AND waiting_tasks_count > 0
-ORDER BY
-waiting_tasks_count DESC
+AND wait_time_ms > 100
```
Comment (glinton) [inline plugins/inputs/sqlserver/sqlserver.go]: Is this intended?

**Comment by m82labs** [comment]
I’ll fix the spacing issues. The last item is intentional yes. It was
unintentional to leave the ORDER BY in there when I first developed that
query.

**Comment by danielnelson** [comment]
@m82labs Are there any README changes we should make with this change?

**Comment by m82labs** [comment]
I can add some verbiage about Azure Managed Instance support. That’s
probably one of the biggest changes.

**Comment by danielnelson** [comment]
Can you check if this resolves 4222?

**Comment by m82labs** [comment]
@danielnelson I *think* it does. I don't have anything to test it against, but I know when this was tested against Azure Managed Instances it would not work until the SQL package was changed.

**Comment by danielnelson** [comment]
@m82labs One last question, all the changes are backwards compatible?  New fields and tags are okay but are there any renames or type changes?

**Comment by m82labs** [comment]
I have not run into a case of this. I am currently running a build
including these changes in a dev environment. One change that could affect
people is that there are cases where more instances have been included for
a given counter, not just the total.

**Comment by danielnelson** [comment]
That should be okay since they everything would be tagged.

**Comment by denzilribeiro** [comment]
It is incrementally a little more data collected but then also provides much deeper insights

Sent from mobile device

On Sep 11, 2018, at 8:41 PM, Mark Wilkinson - m82labs <notifications@github.com<mailto:notifications@github.com>> wrote:

I have not run into a case of this. I am currently running a build
including these changes in a dev environment. One change that could affect
people is that there are cases where more instances have been included for
a given counter, not just the total.

**Comment by danielnelson** [comment]
Users always have the option to trim out unwanted data with tagpass, tagdrop, fieldpass, etc so thats no problem.

Thanks again @m82labs and @denzilribeiro

**Comment by erik-wramner** [comment]
The readme for the plugin specifies grant VIEW SERVER STATE and VIEW ANY DEFINITION and that doesn't work on Azure. Can you please share the privileges you granted to the telegraf account? Or ideally update the readme for the plugin.

**Comment by m82labs** [comment]
@erik-wramner can you try granting VIEW DATABASE STATE. I'll try to get something set up to test today, but that might be the permission needed.

**Comment by erik-wramner** [comment]
I would like to reopen this issue, please. Perhaps it is only documentation, but the plugin is still not working. I tried connecting directly to a database with a user with VIEW DATABASE STATE and the plugin crashes. First it complains about permissions:

`VIEW SERVER STATE permission was denied on object 'server', database 'master'.`

Then it crashes:

`panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x0 pc=0x1226554]

goroutine 24 [running]:
github.com/influxdata/telegraf/plugins/inputs/sqlserver.(*SQLServer).accRow(0xc04240c400, 0x1fd9b51, 0x265, 0x1, 0xc0437ca180, 0xc, 0xc, 0x21b0580, 0xc0421f8860, 0x218bf80, ...)
        /go/src/github.com/influxdata/telegraf/plugins/inputs/sqlserver/sqlserver.go:222 +0x754
github.com/influxdata/telegraf/plugins/inputs/sqlserver.(*SQLServer).gatherServer(0xc04240c400, 0xc04228c161, 0x9e, 0x1fd9b51, 0x265, 0x1, 0xc0437ca180, 0xc, 0xc, 0x21b0580, ...)
        /go/src/github.com/influxdata/telegraf/plugins/inputs/sqlserver/sqlserver.go:178 +0x220
github.com/influxdata/telegraf/plugins/inputs/sqlserver.(*SQLServer).Gather.func1(0xc042039490, 0x21b0580, 0xc0421f8860, 0xc04240c400, 0xc04228c161, 0x9e, 0x1fd9b51, 0x265, 0x1, 0x0, ...)
        /go/src/github.com/influxdata/telegraf/plugins/inputs/sqlserver/sqlserver.go:141 +0xb7
created by github.com/influxdata/telegraf/plugins/inputs/sqlserver.(*SQLServer).Gather
        /go/src/github.com/influxdata/telegraf/plugins/inputs/sqlserver/sqlserver.go:139 +0x1da`

This is with the build telegraf-1.8.0~rc2_windows_amd64 which should contain the fix?

Please update the documentation or at least this issue with a working connection string for a user with well-defined permissions and (if necessary) what edition of SQL Database that is supported. I'm not using managed instance, which is still in preview.

**Comment by danielnelson** [comment]
@erik-wramner Can you open the panic as a new issue?

**Comment by erik-wramner** [comment]
The stacks are the same. Is it possible that telegraf-1.8.0~rc2_windows_amd64 contains the old plugin? It is a new build, only a few days old, but it is odd that the line numbers are the same?

**Comment by danielnelson** [comment]
@erik-wramner I don't think its possible, the changes in 1.8 mostly just modify the queries later in the file.

**Pull request merged**